### DISCUSSION TITLE
Call for Feedback

### DISCUSSION THREAD

**Discussion by jeremyong-az** [o3de/community#103] [General]
The Technical Steering Committee (the TSC) would like to solicit _your_ feedback to help guide the development and processes used by O3DE. Please reply to any of the following questions, and state the SIG(s) you represent or participate in.

1. For non-realtime communication (i.e. not Discord), what is your preferred medium of communication and why? (between Github discussions and email lists)
2. Do you find the PR process to be satisfactory? Are there things you would like to see changed?
3. Do you find the RFC process to be satisfactory? Are there things you would like to see changed?

**Comment by ghost** [comment]
Hello!

1. I personally would like to see a community to be built at [stack exchange](https://stackoverflow.com/teams/pricing). It has good search functions and indexing, great forum and has some [affordable options](https://stackoverflowsolutions.com/teams-free?utm_source=so-owned&utm_medium=post&utm_campaign=campaign-28&utm_content=post-2&_ga=2.232125599.1890273754.1636996187-1175058269.1635266471). Their [Area 51](https://area51.stackexchange.com/) demo community page shows how the basic functionality of questions that are being upvoted/downvoted and marked as solutions during the progress of a thread.

2. Yes, I like the idea that everything is transparent and people take their time to answer questions and are very helpful. That is crucial to build a strong community and with a good fanbase.

3. No experience there yet, I'm meeting the graphics and content guys on Wednesday for the first time (I'll edit this comment after that meeting) :)


Sincerely,
Niklas Henricson
Epos Studios LLC

**Reply by jeremyong-az** [reply]
I could see stack exchange being useful for the Q&A section for helping users/dev of the engine, but maybe less so for future proposal discussion or debates about existing/future design and implementation. For the latter topics, do you have a preference for communication medium, or would you also prefer SE for that as well?

**Reply by ghost** [reply]
> I could see stack exchange being useful for the Q&A section for helping users/dev of the engine, but maybe less so for future proposal discussion or debates about existing/future design and implementation. For the latter topics, do you have a preference for communication medium, or would you also prefer SE for that as well?

You're right. SE is best suited for technical issues, while Github already provides a good platform for discussion. If you think you can host something elsewhere, then phpBB forums have a really good engine to host all that including moderation/administration, direct messages, etc and they've been around forever.

**Reply by thefranke** [reply]
Regarding SE: I'd like to point out that we also have the Github Discussion page [with a Q&A tag/subsection](https://github.com/o3de/o3de/discussions/categories/q-a). It's by far not as visible as SE though.

**Reply by olegbc1** [reply]
> 
> > I could see stack exchange being useful for the Q&A section for helping users/dev of the engine, but maybe less so for future proposal discussion or debates about existing/future design and implementation. For the latter topics, do you have a preference for communication medium, or would you also prefer SE for that as well?
> 
> You're right. SE is best suited for technical issues

I am sorry, I don't think that is true at all. GitHub Issues can be linked to pull requests and vice versa. That alone is probably enough of a reason to stick with it. I do agree with @jeremyong-az on Stack Exchange being a potential site Q&A-type inquiries.

**Comment by sptramer** [comment]
My only suggestion is regarding (2) and (3): We need better synchronization points where it's understood when documentation is needed (or should be made available) as items go through PR. This is a larger issue outside of the current process regarding where docs contributors can set checkpoints for work to be started/in-flight/complete, and contribute to the definition of **DONE** for features.

**Reply by jeremyong-az** [reply]
(putting my user hat on for a second) are there docs currently for writing and contributing docs available I could see? Is this something that may warrant earlier prioritization so that more people can contribute to the docs?

**Reply by sptramer** [reply]
Just like for code contribs, [we have a whole section on docs contribs](https://o3de.org/docs/contributing/to-docs/). Right now we don't have any good understanding of where these checkpoints happen and where they're appropriate, which is why I think it's a good discussion to have between docs and the TSC.

**Reply by olegbc1** [reply]
@kberg-amzn introduced me to doxygen as he uses it to autogen [AzNetworking documentation](https://o3de.org/docs/api/frameworks/aznetworking/index.html). I can't imagine mandating it across the entire codebase though...

**Comment by jeremyong-az** [comment]
I will jot down my own personal opinions (feel free to agree/disagree):

1. I prefer email for discussion (easy to check on phone, one less site to visit, notifications, can use tools available in my email program like marking a message as important, etc.)
2. PR process seems to work well, would like continued emphasis on shortening AR without impacting (or even improving) coverage
3. RFC process in my opinion may be overly biased towards in-person discussions (via Discord voice comm) instead of offline (via email), which makes the ability to give input attendance-based. Also adds a lot of latency to discussing an RFC.

**Reply by monroegm** [reply]
> I prefer email for discussion (easy to check on phone, one less site to visit, notifications, can use tools available in my email program like marking a message as important, etc.)

This may address your requirements. If you are subscribed to a GitHub discussion it will be sent to your email and you can reply from there. You can set it up so you are automatically subscribed to all discussions in a specific repository (like subscribing to a mailing list). Notifications for the same discussion have the same subject line so they can be threaded.

**Reply by ghost** [reply]
> I will jot down my own personal opinions (feel free to agree/disagree):
> 
> 1. I prefer email for discussion (easy to check on phone, one less site to visit, notifications, can use tools available in my email program like marking a message as important, etc.)
> 2. PR process seems to work well, would like continued emphasis on shortening AR without impacting (or even improving) coverage
> 3. RFC process in my opinion may be overly biased towards in-person discussions (via Discord voice comm) instead of offline (via email), which makes the ability to give input attendance-based. Also adds a lot of latency to discussing an RFC.

If you prefer it for personal needs, I have no objection; Overall I'm against emails. They are unreliable (can get stuck on spam filters, become undelivered, delayed, the the sender doesn't know if the receiving side has acknoledged the information). 

Also another reason (unrelated to our voting session here) why I'm against e-mails provide also a false hope that people are more accessible and camp their emails. I tend to look at my emails once a day or every two days. Important emails can thus fall behind so I always ask people with urgest requests to either call me, or text me instead of assuming I'm camping my outlook 24/7.

**Reply by jeremyong-az** [reply]
Yup actually I suggested discussions in a previous meeting but a countersuggestion was that non-tech contributors might find email more accessible. Totally not against GitHub discussions as a viable option either for the reasons you mention.

**Comment by AMZN-Liv** [comment]
1. I prefer GitHub discussions for easy indexing and browsing compared to email, due to the volume that I experience of the latter. 
2. No opinion as I rarely submit PRs. For PRs related to SIG governance, it seems fine.
3. RFC process - I agree that it is more heavily biased towards synchronous conversations to async ones. I think we are still also experiencing some places where it is not clear whether the scope of change/proposal necessitates an RFC, but that the process itself generally is satisfactory.

**Comment by thefranke** [comment]
1. I prefer Github discussions, my mailbox is already pretty stacked and its hard to separate discussions properly.
2. So far so good. I tend to forget to sign-off my commits, but that's all.
3. No opinion here other than that the synchronization points will sooner or later really make everything crawl, given that we only have so much discussion time.

**Comment by lmbr-pip** [comment]
Personal views (have added as a discussion item for next SIG-Network meeting to get official SIG response, if thats required).

1. I think GitHub discussions are easier and more open than email. Puts users in charge and you have more control over the content you want to follow.

2. PR process is mostly working. However, I do not see extensive use of drafts for early engagement, I do see inconsistencies esp on testing (either in description or unit/integration test changes that come with changes), makes it very hard to see whats actually been tested. I also see very few references to RFC in large changes which makes it hard to go back and find technical details.  New process of ensuring PRs are assigned to SIGs seems to be working great.

3. RFCs: We should have a central notification point for each SIG and across O3DE for RFCs as it all seems very siloed inside a SIG. Concern that SIGs are all following their own path for archiving, which is going to be confusing in the future (no consistent naming etc). Do not have clear guidance to ensure that cross SIG RFCs are accepted by all impacted SIGs. We are also not archiving the RFCS centrally (no use of o3de/rfcs) which is confusing, we either use this repo or we drop it.

**Comment by Ulrick28** [comment]
I prefer GitHub discussions instead of email. 

- Many corporate organizations have limits on how long sent and received emails can be stored, whereas discussions is a permanent archive. 
- Conversations are much easier to track via Discussions than via email
- It's easier to tell which discussions are gaining traction and which are not, how much contribution there is, and how long the discussions has been around via the Discussions board. Email is extremely difficult to gain these sorts of insights.
- It is much easier to track when and what decisions are made via the Discussions board than via email. This is important for paper trails or when there is a later dissenting opinion. 
- Potential drawback of Discussions board: Discussions is visible via the phone, but not as intuitive as email.

**Reply by monroegm** [reply]
I'm a fan of GitHub discussions as well, but I'm not sure if you're considering that email lists also have a web interface. Example: https://lists.o3de.org/g/main/topics for main@lists.o3de.org

**Reply by Ulrick28** [reply]
Good to know, still doesn't change my vote :)

**Comment by caniszczyk** [comment]
+1 to GitHub discussions

On Tue, Nov 16, 2021 at 11:27 AM Gavin Monroe ***@***.***>
wrote:

> I'm a fan of GitHub discussions as well, but I'm not sure if you're
> considering that email lists also have a web interface. Example:
> https://lists.o3de.org/g/main/topics for ***@***.***
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/o3de/community/discussions/103#discussioncomment-1653670>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAPSII5XSTWHVWJMM4NPP3UMKWCRANCNFSM5HWAXNJA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>
-- 
Cheers,

Chris Aniszczyk
https://aniszczyk.org

**Comment by monroegm** [comment]
1. +1 to GitHub discussions.
2. I think the PR process is generally working. I would like to see PRs more consistently being linked back to an associated GitHub issue (if one exists). 
3. I think we should only put an RFC into a SIG-specific repo if it is a proposed change to the SIG itself or any SIG-specific artifacts. All other RFCs should go to a central location where they can be easily discovered by everyone. This would also make it easier to create a Discord bot command for listing all of the open RFCs.

**Reply by Ulrick28** [reply]
+1 to the central repository for RFC's. The sig repos do not get a lot of traffic.

**Comment by antonmic** [comment]
2. PR process is pretty good when AR works and the PR is medium to long. Unfortunately AR still has seemingly random failures which can be quite frustrating, and sometimes fails for reasons that are hard to call blocking (saw someone's AR fail not too long ago because a piece of the code had tabs instead of spaces, we really shouldn't be blocking submission and forcing engineers to re-run AR for something like this). On the review side, I would still love a bypass button for tiny changes like fixing typos in comments and tweaking default values etc, there's no reason for engineers to have to randomize two other engineers and run AR if they're just updating a comment or making some extremely trivial change.
No strong feelings about 1. or 3.

**Comment by lmbr-pip** [comment]
I asked the TSC (because of concerns raised by others external to this thrad) to get some guidance on RFC archiving and how we should utilize [o3de/rfcs](https://github.com/o3de/rfcs) at the [12/7 meeting]( https://github.com/o3de/tsc/issues/3). Unfortunately TSC did not have time to discuss so has moved to next meeting (ETA 12/14).

Have concerns also around low engagement around RFCs in SIG. Many do not attract comment or discussion. Unsure if low engagement is part of the start-up process, or if we are missing learnings from other open-source communities.

**Comment by amznestebanpapp** [comment]
1. I prefer to keep everything in GitHub. Things to improve:
- its hard to filter the questions/ideas/etc that need feedback. I would suggest to create some labels to track what was already addressed (e.g. if it is an idea maybe label it as "accepted"). 
- be more aggressive on marking things as "solved". If they are not really solved, the author can comment and the tag removed. I find myself re-reading and going through the same questions over and over. Most of the times the author just didnt reply which likely indicates that they are not running into the issue anymore.

2. PR is satisfactory. Things to improve:
- would like a bot to trigger dco check (sometimes dco doesnt trigger, like when the base branch is changed)
- would like a bot to trigger jenkins (currently have to go and find the job in jenkins)
- would like more cleanup on PRs. Some PRs take a long time to go through or get closed. We need to increase the amount of active reviewers and establish rules on how long a PR can stay open without feedback from the author. I would like to see an auto-close if the author has not addressed feedback after X amount of time.

3. No comments yet, I need to go through it a couple of times to give some feedback.

**Comment by hultonha** [comment]
_1. For non-realtime communication (i.e. not Discord), what is your preferred medium of communication and why? (between Github discussions and email lists)_

- I would prefer to keep everything in GitHub. It makes it much easier to know what is the source of truth and where to go looking for things.
- It would be great to take advantage of GitHub features such as Discussions as well as Issues to spark conversations and discuss changes/improvements.

_2. Do you find the PR process to be satisfactory? Are there things you would like to see changed?_

- I think it's pretty good on the whole, but I'd definitely be in favour of reviewers being much stricter about tests. All contributors should be thinking about tests and making sure coverage is added for as many things as possible. The coverage doesn't have to be exhaustive but it's super important to have something that ensures the feature you've added won't break if someone else touches it.
  - The mentality should be if contributor A changes some code, and it inadvertently breaks contributor B's feature/system, it is contributor B's fault for not having a test in place to catch the breakage, not contributor A.
  - This of course would need some flexibility (especially for legacy systems) but that will help encourage tests for PRs as a matter of course. We should all be doing this to set a good example that this is the standard we want.

_3. Do you find the RFC process to be satisfactory? Are there things you would like to see changed?_

- To be brutally honest right now it seems like a box ticking exercise, as there just isn't enough engagement with the RFCs themselves. They often get posted, sit there for a while, and then development on them starts as no one has voiced opposition (reminds me a little of the Hitchhikers Guide to the Galaxy demolition notice 😉). We likely should persevere with them, they do have a place, but right now it doesn't feel like they're having the intended effect.
